langchain/src/prompts/base.ts

```diff
@@ -35,14 +38,35 @@ export abstract class BasePromptTemplate implements BasePromptTemplateInput {
 
   outputParser?: BaseOutputParser;
 
+  partialVariables?: InputValues;
+
   constructor(input: BasePromptTemplateInput) {
-    const { inputVariables } = input;
+    const { inputVariables, partialVariables } = input;
     if (inputVariables.includes("stop")) {
       throw new Error(
         "Cannot have an input variable named 'stop', as it is used internally, please rename."
       );
     }
     Object.assign(this, input);
+    Object.assign(this, partialVariables);
+  }
+
+  abstract partial(values: InputValues): Promise<BasePromptTemplate>;
+
+  mergePartialAndUserVariables(userVariables: InputValues): InputValues {
+    const partialVariables = this.partialVariables ?? {};
+    const partialKwargs: InputValues = {};
+    for (let i = 0; i < Object.keys(partialVariables).length; i += 1) {
+      const key = Object.keys(partialVariables)[i];
+      const value = partialVariables[key];
+      if (typeof value === "string") {
+        partialKwargs[key] = value;
+      } else {
+        partialKwargs[key] = value();
```

Comment:
1) `partialKwargs` is pretty python-centric terminology. Probs better to rename to `partialValues`
2) We probably want to think about the typing for `inputValues` here and should support `value` being either an async function or string. Right now, if value is async, `partialKwargs[key]` would be assigned a promise which isn't what we want.